# Домашнее задание 1

In [1]:
import numpy as np

In [2]:
y = [45, 55, 50, 59, 65, 35, 75, 80, 50, 60]
X = np.array([[1, 1, 1, 1, 1, 1, 1, 1,  1, 1],
              [1, 1, 2, 1, 3, 0, 5, 10, 1, 2]])

In [3]:
def calc_mse(y_pred, y):
    return np.mean((y - y_pred) ** 2)

1. Подберите скорость обучения (alpha) и количество итераций:

In [4]:
n = X.shape[1]
alpha = 1e-2

Для того, чтобы найти лучшую скорость обучения и количество итераций, создадим цикл, в котором и будем определять необходимое количество итераций для данной скорости обучения (_alpha_). В качестве критерия остановки будем использовать целевую разницу между предыдущим и текущим значением `mse`.

In [5]:
W = np.array([1, 0.5])
target_diff = 1e-7
diff = np.inf
i = 0
while target_diff < diff:
    y_pred = np.dot(W, X)
    err = calc_mse(y, y_pred)
    for ii in range(W.shape[0]):
        W[ii] -= alpha * (1/n * 2 * np.sum(X[ii] * (y_pred - y)))
    new_err = calc_mse(y, np.dot(W, X))
    diff = err - new_err
    if i % 100 == 0:
        print(i, W, err, diff)
    i += 1
print(i, W, err, diff)

0 [2.102 3.9  ] 3173.15 1087.9666360000006
100 [31.88770806  6.74418155] 175.19445858001853 2.6740020558778497
200 [41.83683774  4.90699865] 61.9177717428135 0.3305917140802208
300 [45.33508261  4.26102097] 47.913169919666785 0.040871651978825696
400 [46.56511152  4.03388672] 46.181755648107604 0.005053036311359449
500 [46.99760587  3.95402334] 45.96769776787538 0.0006247160251007244
600 [47.14967657  3.92594232] 45.941233404700036 7.72347729096623e-05
700 [47.20314662  3.91606866] 45.93796156758051 9.548674803738777e-06
800 [47.2219474   3.91259695] 45.93755706443538 1.1805199164882652e-06
900 [47.228558    3.91137626] 45.937507054979434 1.4594981934124007e-07
920 [47.22920369  3.91125703] 45.93750474242528 9.810889878281159e-08


Таким образом подобраны оптимальные значения c заданной точностью `mse`:
* скорость обучения `alpha` = 1e-2;
* количество итераций - 920. Округлим в большую сторону - 1000.

In [6]:
# подставим найденные значения в заданный цикл
W = np.array([1, 0.5])
for i in range(1000):
    y_pred = np.dot(W, X)
    err = calc_mse(y, y_pred)
    for ii in range(W.shape[0]):
        W[ii] -= alpha * (1/n * 2 * np.sum(X[ii] * (y_pred - y)))
    if i % 100 == 0:
        print(i, W, err)
print(i, W, err)

0 [2.102 3.9  ] 3173.15
100 [31.88770806  6.74418155] 175.19445858001853
200 [41.83683774  4.90699865] 61.9177717428135
300 [45.33508261  4.26102097] 47.913169919666785
400 [46.56511152  4.03388672] 46.181755648107604
500 [46.99760587  3.95402334] 45.96769776787538
600 [47.14967657  3.92594232] 45.941233404700036
700 [47.20314662  3.91606866] 45.93796156758051
800 [47.2219474   3.91259695] 45.93755706443538
900 [47.228558    3.91137626] 45.937507054979434
999 [47.23086913  3.91094949] 45.93750089064509


*2. В этом коде мы избавляемся от итераций по весам, но тут есть ошибка, исправьте ее:

In [7]:
W = np.array([1, 0.5])
for i in range(1000):
    y_pred = np.dot(W, X)
    err = calc_mse(y, y_pred)
    '''for ii in range(W.shape[0]):
    W[ii] -= alpha * (1/n * 2 * np.sum(X[ii] * (y_pred - y)))'''
    # W -= (alpha * (1/n * 2 * np.sum(X * (y_pred - y))))
    W -= alpha * (1/n * 2 * X.dot(y_pred - y))
    if i % 100 == 0:
        print(i, W, err)

0 [2.102 3.9  ] 3173.15
100 [31.88770806  6.74418155] 175.19445858001842
200 [41.83683774  4.90699865] 61.9177717428135
300 [45.33508261  4.26102097] 47.913169919666785
400 [46.56511152  4.03388672] 46.181755648107604
500 [46.99760587  3.95402334] 45.96769776787538
600 [47.14967657  3.92594232] 45.941233404700036
700 [47.20314662  3.91606866] 45.93796156758051
800 [47.2219474   3.91259695] 45.93755706443538
900 [47.228558    3.91137626] 45.937507054979434
